src/HeatmapView.js

```diff
@@ -0,0 +1,181 @@
+  constructor(props) {
+    super(props)
+
+    this.state = {
+      data: [],
+      filteredData: [],
+      selectedClusterId: null,
+      isLoading: true,
+      hasError: null
+    }
+  }
+
+  async _fetchAndSetState({resource, host}) {
+    this.setState({
+      isLoading: true
+    })
+
+    const url = URI(resource, host).toString()
+
+    try {
+      const response = await fetch(url)
+
+      if (!response.ok) {
+        throw new Error(`${url} => ${response.status}`)
+      }
+
+      const jsonData = await response.json()
+      this.setState({
+        data: await jsonData,
+        filteredData: await jsonData,
+        isLoading: false,
+        hasError: null
+      })
+    } catch(e) {
+      this.setState({
+        data: null,
+        isLoading: false,
+        hasError: {
+          description: `There was a problem communicating with the server. Please try again later.`,
+          name: e.name,
+          message: e.message
+        }
+      })
+    }
+  }
+
+  componentDidMount() {
+    this._fetchAndSetState(this.props)
+  }
+
+  componentDidUpdate(previousProps) {
+    if (previousProps.resource !== this.props.resource) {
+      this._fetchAndSetState(this.props)
+    }
+  }
+
+  componentDidCatch(error, info) {
+    this.setState({
+      hasError: {
+        description: `There was a problem rendering this component.`,
+        name: error.name,
+        message: `${error.message} – ${info}`
+      }
+    })
+  }
+
+  render() {
+    const { data, filteredData, selectedClusterId, isLoading, hasError } = this.state
+    const { wrapperClassName, plotWrapperClassName } = this.props
+    const { ks, ksWithMarkers, selectedK, onSelectK } = this.props
+    const { hasDynamicHeight, defaultHeatmapHeight, heatmapRowHeight } = this.props
+
+    const kOptions = ks.sort((a, b) => a-b).map((k) => ({
+      value: k.toString(),
+      label: `k = ${k}`,
+      isDisabled: ksWithMarkers ? !ksWithMarkers.includes(k) : false
+    }))
+
+    const allClusterIds = Array.from(Array(Number.parseInt(selectedK)+1).keys()).slice(1)
+    const clusterIdsWithMarkers = data && _.uniq(data.map(x => x.clusterIdWhereMarker))
+
+    let clusterIdOptions = allClusterIds
+      .sort((a, b) => a-b)
+      .map((clusterId) => ({
+        value: clusterId.toString(),
+        label: `Cluster ${clusterId}`,
+        isDisabled: clusterIdsWithMarkers ? !clusterIdsWithMarkers.includes(clusterId) : false
+      }))
+
+    // Add default "All clusters" option at the start of the options array
+    clusterIdOptions.unshift({
+      value: `all`,
+      label: `All clusters`,
+      isDisabled: false // always show this option
+    })
+
+    return (
+      hasError ?
+        <CalloutAlert error={hasError}/> :
+        <div>
+          <div className={wrapperClassName}>
+            <div className={`small-12 medium-6 columns`}>
+              <PlotSettingsDropdown
+                labelText={`Show marker genes for:`}
```

Comment:
I think the `labelText` for two input boxes can be more specific to differentiate `k` and `Cluster` values.

Reply:
I know it's a bit repetitive, but I couldn't think of anything better that would make sense to the user. Any suggestions?

Reply:
I would say `Select k` if users know what `k` means and `Select cluster`, does it make more sense?

Reply:
How about *Number of clusters* for the first label?

Reply:
Like Number of clusters: k = 8? Or should I remove the "k = " part altogether?